Add export functionality

awalter-bnl opened the pull request:
This is meant to add export functionality to bluesky-browser as outlined in issue #34 

from issue #34 the expected process at the end is:
```
1. User selects search results and clicks Export. This opens up a dialog box.
2. Dialog box contains tabs (maybe tabs along the side?), one tab per format.
3. Each format can be activated on/off and configured (directory, file_prefix, custom options).
4. Click Go.
```

Still a work in progress, comments on what we have so far is appreciated.

So far I have added the new 'export' button, it is visible after selecting a scan from the search results.

currently this automatically exports a file containing the meta-data to /test_data/test_data_'scan_id'

TODO: 
- [x] add dialog box to ask for parameter inputs with 'Go button'
- [x] add additional tabs to dialog box for other suitcase options
- [ ] add on-disk persistence of export parameters for quick saving second time around.
- [ ] add list of required suitcases to configuration file

awalter-bnl commented:
attn @danielballan early feedback is appreciated (but not expected).

awalter-bnl commented:
I have now added the input dialog and the tabs for different file types.  Below are screenshots of the browser with the new 'export' button (just below the copy uid to clipboard button) and the 'export dialog' screen.

As their is already quite a bit of code here I would like to pause and wait for some reviews before finishing of the parts to do with persisting default values to disk and adding the list of suitcases to the configuration file.

@mrakitin @danielballan and @tacaswell your input would be appreciated.

![Screenshot from 2019-06-26 07-46-28](https://user-images.githubusercontent.com/25772709/60178230-fd562980-97e8-11e9-813a-1103969f7124.png)
![Screenshot from 2019-06-26 07-46-56](https://user-images.githubusercontent.com/25772709/60178240-03e4a100-97e9-11e9-94fc-1a31c2fa9cfa.png)

danielballan commented:
665c6d5 looks good. Another design hint/nudge: I think it would work well to follow the extension pattern we have laid out for the RunViewer and FigureManager. Something like:

```
class CSVSuitcase(QWidget):
    # the CSV tab in the export dialog box
    ...

class TIFFSuitcase(QWidget):
    ...
  
class ExportDailogBox:
    suitcases = [CSVSuitcase, TIFFSuitcase, ...]
```

In this way, we will be able to add tabs during the cycle by editing the config file without editing the source of bluesky-browser itself.

danielballan commented:
Just got a look at the screenshots above---very quick work! Nice!

awalter-bnl commented:
> In this way, we will be able to add tabs during the cycle by editing the config file without editing the source of bluesky-browser itself.

This will be possible already with the layout I currently have (as the list of suitcases to include can be added to the configuration). Further my approach means we don't need to add a 'new' tab class to bluesky-browser every time we add a new suitcase, just add it to the list of required suitcases.

@danielballan does this cover your concern effectively?

Edit: Just saw @danielballan's slack post indicating this was meant to be posted prior to the second commit. I think this new commit resolves the issues @danielballan had.

danielballan commented:
Not exactly, but what I'm suggesting is a relatively small code change.

1. If you define `suitcase_list` as a traitlet `export_widget` like this:

   ```python
   from traitlets import List

   class export_widget(ConfiurableQWidget):
       suitcase_list = List(['csv', 'json_metadata'])
       suitcase_default_values = ...
   ```

   Then it will be configurable via:
   ```
   # bluesky_browser_config.py
   c.export_widget.suitcase_list = 

2. Is there any need to have `suitcase_list` _and_ `suitcase_default_values`? It looks like `suitcase_list == list(suitcase_default_values)`. If we just have `suitcase_default_values`, there is less to configure and no possibility of the two lists getting out of sync.

3. You are proposing that all the suitcases can use `suitcase_params_widget` as their class. I think each suitcase will need its own widget (likely a simple subclass of `suitcase_params_widget`) to add inputs for format-specific options like CSV separator and the various TIFF options. For some formats, at least to start, using `suitcase_params_widget` itself may be enough, but if we write a custom subclass anyway (even an empty one) that gives us a place to scope its configuration, which will be important when we get to syncing to workspace state to disk.

awalter-bnl commented:
@danielballan responses to your points are below.

1. So far I have not moved this to a 'configurable item' but that is the plan (see above post describing that I would like feedback on the existing code before extending it).

2. `suitcase_list` is the list of possible suitcases from the configuration file, `suitcase_default_values` is intended to be the 'persist to disk' input parameters. Therefore `suitcase_list` may or may not be equal to `list(suitcase_default_values)`

3.  This requires a more detailed response, and perhaps an in-person discussion, but my short answer is: I would prefer to make the parameters displayed for each suitcase configurable, so that only the amount of complexity that is really needed is presented to the user. So I think this can be done by a kwarg in the standard widget listing what parameters to ask for inputs for.

awalter-bnl commented:
Making a note here, had an in-person discussion with @danielballan. The resolutions are:
1. To resolve this we will make `suitcase_list` a `traitlet` on `export_widget`
2. The difference is acknowledged, @danielballan will work on a mechanism to do the persist to disk part.
3. We have worked out a method using `traitlets` to allow for this to be configurable, but it requires very thin child classes of base `suitcase_params_widget` for each different `suitcase`.

I will work toward this now !-).

awalter-bnl commented:
@danielballan can you please take a look and see if the latest updates are heading in the right direction (it is working currently with `suitcase.csv` and partilaly working with `suitcase.json_metadata`.

If this looks good I will complete the widget classes for all the exisitng non-backend suitcases.

danielballan commented:
The SuitcaseWidget and csvSuitcaseWidget look great—nice work capturing pandas’ many options. The export_widget can be much simplified in a way that I can’t type on a phone. Will elaborate soon.

danielballan commented:
Reaching into `sys.modules` is not something we should ever have to resort unless we are handing a very special case. Instead, I suggest following this pattern:

https://github.com/bluesky/bluesky-browser/blob/9e4ef22024f3f41a8ba845e1f63f53e0273452c3/bluesky_browser/viewer/viewer.py#L210-L212

We might want to shorten the names of the suitcase class to things like `CSVSuitcase` (as opposed to `CSVSuitcaseWidget`) to make the configuration more readable.

danielballan commented:
Notice that this approach allows the user to add classes to that list that might be defined in their own separate packages or typed directly into the configuration file. That's why I took this approach for configuring/extending the `RunViewer` and `FigureManager`, and it applies here just as well.

danielballan commented:
Another quick thought --- we could also do this like a dict:

```py
class ExportWidget(...):
    suitcases = Dict({'csv': CSVSuitcase, 'tiff': TIFFSuitcase', ...}, config=True)
    show = List(['csv'], config=True)
```

The `suitcasees` configurable traitlet would be useful to expert users who wanted to add a new class or replace one of our classes with a custom one (say, that laid out the GUI differently or exposed some clever option not thought of by us).

The `show` configurable traitlet contains plain strings, corresponding to the keys of the `suitcases` `Dict`, which would be much easier for a novice user to edit to add or remove the names of the exporters they are interested in.

Alternatively, we could do it like this:

```py
class CSVSuitcase(SuitcaseWidget):
    suitcase_module = 'csv'  # the class knows the name of the suitcase package it is associated with
    ...

class ExportWidget(...):
    suitcases = List([CSVSuitcase, TIFFSuitcase], config=True)
    show = List(['csv'], config=True)
```

Either seems like a fine solution to me.

awalter-bnl commented:
I have made some changes which allow the use of `traits` for configuration and allow for input widgets other than `QLineEdit`.

Personally I think this is getting pretty messy, with the main issue being the need to configure:
1. which suitcases have included tabs
1. which parameters are included for each tab

The result is that I have had to add 'mapping dictionaries' for each of these. The alternatives I have come up with are also not ideal:
A. keep the configurability but hard-code each of the parameters(suitcases) inside an `if parameter (suitcase) in parameter(suitcase)_list:` block 
B.  remove the configurability and provide a subset of parameters(suitcases) hardcoded into the widgets.

Before continuing on I would like some feedback, take a look at csvSuitcaseWidget for an example of how each of the widgets will look in the current scheme. As @danielballan and I discussed this structure already I would also like some feedback from fresh eyes to see if this is not getting to messy(@mrakitin or @tacaswell ?).

If we have agreement that this approach is ok I will continue to add other suitcases

mrakitin commented:
I like the "mapping dictionaries" approach. Will give my feedback after a more thorough review.

danielballan commented on bluesky_browser/suitcase.py:
```diff
@@ -0,0 +1,298 @@
+            temp_widget = getattr(self, f'{parameter}_widget')
+            # write the default value from the trait to the widget.
+            slot = getattr(temp_widget,
+                           f'set{self._upper(_value_map[widget_type])}')
```
This string-munging here is brittle: what if we encounter a class that doesn't spell things just so? What if instead we subclassed each input class to give it a consistent API:

```python
class MyQLineEdit(QLineEdit):
    def set_default(self, value):
        return self.setText(value)
    def observe(self, callback):
        return self.onTextChanged(callback)
```

That is an approach we can apply consistently to any widget types we later need. It also makes it easier for third parties to introduce new widgets into the system. In the current approach, they would need to have something added to `_value_map`.

There a second compelling reason to take the subclassing approach: the input types from Qt come with some strange defaults, such as the float input type imposing default limits of 0 to 10. In a subclass, we would have the opportunity of overriding those here.

@tacaswell began similar work in mily [right here](https://github.com/NSLS-II/mily/blob/77d12d43baa5e3ff297d417b475a5188c9e5ff70/mily/widgets.py#L54-L110) because these widgets could be useful to PyDM also. I propose putting a PR into mily to add the widget types we need here, and then adding a mily dependency to bluesky-browser.

danielballan commented on bluesky_browser/suitcase.py:
```diff
@@ -0,0 +1,298 @@
+            trait = self.traits()[parameter]
+            widget_type = trait.metadata['widget']
+            # create, and reference, the widget.
+            setattr(self, f'{parameter}_widget', widget_type())
```
A minor thing: this monkey-patching and string-munging stands out to me. Could we instead define `self.widgets = {}` in `__init__` and then use `widget = widget_type(); self.widgets[parameter] = widget` here?

awalter-bnl commented:
Based on @danielballan latest requests I would like to take a step back before proceeding, as I think we are making some decisions that could have large long-term impacts without considering those impacts carefully enough. This is despite me usually being a strong advocate for factory methods to auto-generate classes in order to reduce the amount of code at the expense of readability and simplicity!-)

Before this latest request I was already worried that this code is to abstract and complicated to follow. Based on the existing code I have issues with:
- global dictionaries that provide mapping from strings to objects.
- many different, strongly disconnected, parts that need updating whenever a new suitcase is added. (the suitcase widget itself, the dictionary mapping `Qwidgets` to values, and the dictionary mapping string names to suitcase classes).
- the large number of sub-classed `Qwidgets` that I have had to generate just to get the `csvSuitcaseWidget`, let alone some of the others. 

@danielballan's latest suggestion now moves part of what needs updating to a completely different repo, and introduces our own sub-class of `Qwidgets` and a custom API.

This is all necessary in order to introduce a significant amount of configuration from a configuration file. Before we head down this path I think we need to discuss, in person (or bluejeans) as a group, the following points:
- do we really need this configurability in a config file, or should/could we obtain this by sub-classing the `xyzSuitcaseWidgets` on a case by case basis when extra parameters and/or suitcases are required?
- if we do want to head down the path of configuration from a file, and start to introduce a custom API (as suggested above) we need to discuss how this should look a little before doing this. This is essentially doing the same job that `ophyd` was designed to do for `bluesky` (i.e. ensure a consistent API from a range of items, in this case `Qwidgets` not `IOC's`) to simplify the code that interacts with those items.

Both of these decisions I think need careful thought and discussion before proceeding.

danielballan commented:
To make sure we are on the same page:

> global dictionaries that provide mapping from strings to objects.

My latest suggestion would remove that entirely.

> many different, strongly disconnected, parts that need updating whenever a new suitcase is added. (the suitcase widget itself, the dictionary mapping Qwidgets to values, and the dictionary mapping string names to suitcase classes).

The dictionary mapping Qwidgets to values goes away, and the dictionary mapping string names to suitcase classes can also go away if we adopt my alternate proposal [in this comment](https://github.com/bluesky/bluesky-browser/pull/35#issuecomment-506060177).

> the large number of sub-classed Qwidgets that I have had to generate just to get the csvSuitcaseWidget, let alone some of the others.

I take the sign that we need these in other libraries, not just bluesky-browser, to be a sign that they are a good idea. I notice that other GUI libraries tend to have consistent names, so this customization brings PyQt's design into line with a practice that we see elsewhere.

> @danielballan's latest suggestion now moves part of what needs updating to a completely different repo, and introduces our own sub-class of Qwidgets and a custom API.

We can always define extra ones in bluesky-browser or even right in the config file if we can't wait for synchronized updates from mily. But this seems like a good opportunity to actually use mily for something.

danielballan commented:
On a call, @awalter-bnl have agreed to back out of attempting automated GUI generation for these. All of the proposed approaches, including my latest one, reduce the amount of boilerplate code but do increase the conceptual complexity.

We can always change our minds about this later, but I think it's likely this is the right long-term course to take.

danielballan commented on bluesky_browser/suitcase.py:
```diff
@@ -0,0 +1,558 @@
+        widget_type = QLineEdit  # The QT widget for this trait
+        slot_name = 'setText'  # The slot name for this trait
+        signal_name = 'textChanged'  # The signal name for this trait
+        # reference the trait and create the trait widget
```
Have you considered making this into a method that takes the four variables above as arguments?

awalter-bnl replied:
@danielballan I have considered it, but am still a little on the fence.

I assume the method would be in the base ``SuitcaseWidget`` class and look something like.

```
def _create_trait_widget(self, trait_name, widget_type, slot_name, signal_name):
    # reference the trait and create the trait widget
    trait = getattr(self, trait_name)
    setattr(self, f'{trait_name}_widget', widget_type())
    trait_widget = getattr(self, f'{trait_name}_widget')
    # write the value of the trait to the widget
    if trait is not None:  # ignore if it is "None"
    getattr(trait_widget, slot_name)(trait)
    # Connect the output of the widget to the file_prefix trait.
    getattr(trait_widget, signal_name).connect(
        lambda output: setattr(self, trait_name, output))
```
And we would then do a series of definitions in each of the child classes like:
```
self._create_trait_widget('directory', QLineEdit, 'setText' 'textChanged' )
self._create_trait_widget('file_prefix', QLineEdit, 'setText' 'textChanged' )
...
```

I am not against this, it does improve readability. I suppose my only (not strongly held) concern is that we are starting to head toward conceptual complexity again.

Is this what you had in mind?

danielballan replied:
Yes. I think the benefits justify the new abstraction in this case.

awalter-bnl replied:
OK I have made this change.

awalter-bnl commented:
sorry for the delay on this latest round of updates, there was a few new issues that I found when working on the new approach and a few suggested tactics that did not work as expected. 

Once again I have ``csvSuitcaseWidget`` fully fleshed out and working and a stripped back ``json_metadataSuitcaseWidget``. Once agreement has been made that this is a good path forward then I will continue on to flesh out the rest of the suticases.

awalter-bnl commented:
So I have made the change to a ``_create_trait_widget()`` method as requested by @danielballan. In addition I have made the following changes in the latest commit:
1. I have updated the ``show_parameters`` methods in both ``suitcaseWidget`` and ``csvSuitcaseWidget`` to ensure that the parents class ``self.parameters`` supercedes the child classes version.
1. I have moved ``self.geometry`` from an initialization kwarg to a trait in both ``suitcaseWidget`` and ``export_widget`` to allow the location and size to be configurable.

~~I am having an issue where there is some persistence of traits occuring. For instance, at one point in the command line I changed the value of ``csvSuitcaseWidget.parameters`` to exclude `'seperator'` from the list. Prior to this everything worked as expected, now `seperator` is no longer showing up on instantiation in the parameters list, despite being in the default value. This issue goes away when I swap this from a trait to a normal list. I will continue to look into where this persistence of a changed value is coming from, but if you have any ideas @danielballan I would be greatful.~~
UPDATED: it appears that when setting ``self.parameters`` as a trait there is a weird connection between ``self.parameters`` and ``super().parameters``, where both take the value of the first one called. setting either one to a list resolves this issue but I am looking for a better solution that keeps them both as traits.

awalter-bnl commented:
Added a fix to the issue crossed out above, will wait for feedback before continuing further.

awalter-bnl commented:
Despite indicating that I was waiting for feedback, during testing I found a potentially very confusing issue which I have resolved. Essentially pressing the 'export' button without having any any of the 'enable exporter' checkboxes checked would produce behavior identical to having something exported without actually exporting anything. I have hence added a warning `QMessageBox`, where 'ok' must be clicked by the user, to warn that no exporting has occured (see screenshot below). I will now wait for feedback before continuing.

![Screenshot from 2019-07-11 07-24-13](https://user-images.githubusercontent.com/25772709/61047486-9d957c00-a3ad-11e9-8153-c6072b968a0a.png)

danielballan commented:
I think this is a misconception about class inheritance at place here. Nothing to do with traitlets specifically. Unless I am very confused, `super().parameters` and `self.parameters` refer to the very same object. Here is a demonstration. Keep is mind that `is` does `id(left_side) == id(right_side)`, so it is checking that the objects are the same, not that their contents are equal.

```python
In [2]: from traitlets import HasTraits, List

In [7]: class A(HasTraits):
   ...:     parameters = List(['a', 'b', 'c'])
   ...:     vanilla_class_attribute = 'thing'
   ...:     def some_method(self):
   ...:         ...
   ...:         
   ...:     

In [10]: class B(A):
    ...:     def test(self):
    ...:         print('parameters', self.parameters is super().parameters)
    ...:         print('vanilla_class_attribute', self.vanilla_class_attribute is super().vanilla_class_attribute)
    ...:         print('some_method', self.some_method is super().some_method)
    ...:         

In [11]: b = B()

In [12]: b.test()
parameters True
vanilla_class_attribute True
some_method False
```

awalter-bnl commented:
@danielballan  this discussion is academic now as I have found a resolution, but FYI your example above misses a crucial point which makes a difference. If you want to discuss this in person I am happy to but I don't think it gains us anything from discussing here in the git issue as this is now a fully working example.

danielballan commented:
Fair enough! Glad you found a solution.

danielballan commented on bluesky_browser/suitcase.py:
```diff
@@ -0,0 +1,389 @@
+                                  'textChanged')
+
+    def show_parameters(self, parameters=[]):
+        '''Walks through ``self.parameters`` then ``parameters`` showing the
```
Docstring needs updating.

danielballan commented on bluesky_browser/suitcase.py:
```diff
@@ -0,0 +1,389 @@
+
+        # reference the trait and create the trait widget
+        trait = getattr(self, trait_name)
+        setattr(self, f'{trait_name}_widget', widget_type())
```
How about defining `self._widgets = {}` in `__init__` and then doing `self._widgets[trait_name] = widget_type()`? That would avoid monkey-patching and string-munging.

danielballan commented on bluesky_browser/suitcase.py:
```diff
@@ -0,0 +1,389 @@
+        '''Walks through ``super().parameters``, ``self.parameters`` then
+        ``parameters`` showing the widgets in each.
+
+        NOTE: This modified version is required for all child classes. It
```
It seemed to me there ought to be a better way than copying this into every child class. I pulled up [the traitlets API documentation](https://traitlets.readthedocs.io/en/stable/api.html#dynamic-default-values) and found that we are not the first people to need this kind of customization.

```python
In [16]: from traitlets import HasTraits, List, default

In [17]: class A(HasTraits):
    ...:     parameters = List()
    ...:     
    ...:     @default('parameters')
    ...:     def _get_default_parameters(self):
    ...:         return ['a', 'b']
    ...: 
    ...:         

In [18]: class B(A):
    ...:     @default('parameters')
    ...:     def _get_default_parameters(self):
    ...:         ret = super()._get_default_parameters()
    ...:         ret.extend(['c', 'd'])
    ...:         return ret
    ...:     

In [19]: a = A()

In [20]: a.parameters
Out[20]: ['a', 'b']

In [21]: b = B()

In [22]: b.parameters
Out[22]: ['a', 'b', 'c', 'd']
```

awalter-bnl commented:
@klauer based on your confused emoji you would like some more info. for everyones information I moved that discussion to a private PM in slack as there @danielballan and I went back and forth with about 5 messages for us both to be on the same page. Doing this on a Git issue would have taken longer, and filled up the git history with information that is not really relevant to the current code. 

I include a summary of the slack interaction below so that others can understand why I found the traitlet behaviour confusing. First `traitlet` attributes do not act in the same way as `class` attributes even though they are defined in essentially the same way.
```
In [1]: from traitlets import HasTraits, List                                                                                  

In [2]: class A(HasTraits): 
   ...:     class_attr = ['a'] 
   ...:     traitlet_attr = List(['a']) 
   ...:      
   ...: class B(A): 
   ...:     class_attr = ['b'] 
   ...:     traitlet_attr = List(['b']) 
   ...:      
   ...:     def test(self): 
   ...:        print('traitlet_attr', self.traitlet_attr is super().traitlet_attr) 
   ...:        print('class_attr', self.class_attr is super().class_attr) 
   ...:  
   ...: b=B()        
   ...: b.test()                                                                                                               
traitlet_attr True
class_attr False
```
Second the default value attributed to the `traitlet` attribute depends on which one was called first.
```
In [3]: class A(HasTraits): 
   ...:     class_attr = ['a'] 
   ...:     traitlet_attr = List(['a']) 
   ...:      
   ...: class B(A): 
   ...:     class_attr = ['b'] 
   ...:     traitlet_attr = List(['b']) 
   ...:      
   ...:     def test(self): 
   ...:        print(f'traitlet_attr: super:{super().traitlet_attr}, self:  {self.traitlet_attr}') 
   ...:  
   ...: b=B()        
   ...: b.test()                                                                                                               
traitlet_attr: super:['a'], self:  ['a']

In [4]: class A(HasTraits): 
   ...:     class_attr = ['a'] 
   ...:     traitlet_attr = List(['a']) 
   ...:      
   ...: class B(A): 
   ...:     class_attr = ['b'] 
   ...:     traitlet_attr = List(['b']) 
   ...:      
   ...:     def test(self): 
   ...:        print(f'traitlet_attr: self:{self.traitlet_attr}, super():  {super().traitlet_attr}') 
   ...:  
   ...: b=B()        
   ...: b.test()                                                                                                               
traitlet_attr: self:['b'], super():  ['b']
```
Both of these I behaviors I found surprising and confusing, but the solution is to not user `super().traitlet_attr` in the same way that one would use `super().class_attr`.

klauer commented:
Thanks, @awalter-bnl, for clarifying! That's something I didn't know about traitlets.

danielballan commented:
Respecting @awalter-bnl's request to step back from this work, @jklynch is taking over in a new PR, #46.
